From 0cc43961b99c05dadd652d02459fe81629434b09 Mon Sep 17 00:00:00 2001
From: Martin Kaiser <martin.kaiser@codasip.com>
Date: Wed, 12 Feb 2025 14:53:51 +0100
Subject: [PATCH] vi: fix cheri crash when opening an empty file

touch ./empty && vi ./empty

crashes vi on a cheri system.

The count_lines function gets two pointers for the upper and lower bounds
and returns the number of lines in this range. The two pointers should
point to the editor buffer (between the global "text" and "end" pointers).

count_lines seems to be optimized for speed and has no range checks for
its parameters. Some callers set one of the bounds to the byte before the
edit buffer. On cheri systems, this causes a bounds violation when this
byte is read.

This fault happens at startup when the buffer is empty and on other
occassions. Add a simple check to count_lines to fix it. The overhead
in count_lines and the increase in code size should be neglectable.
---
 editors/vi.c | 4 ++++
 1 file changed, 4 insertions(+)

Upstream-Status: Inappropriate [CHERI specific]

diff --git a/editors/vi.c b/editors/vi.c
index 56fab6783..b4458f40e 100644
--- a/editors/vi.c
+++ b/editors/vi.c
@@ -690,6 +690,10 @@ static int count_lines(char *start, char *stop)
 		start = stop;
 		stop = q;
 	}
+	// avoid reading outside the buffer (from text to end)
+	if (start < text) {
+		return 0;
+	}
 	cnt = 0;
 	stop = end_line(stop);
 	while (start <= stop && start <= end - 1) {
-- 
2.43.5

